File: src/ios/AssetManifest.swift
struct AssetManifest {
  struct Entry {
    let filePath: String
    let URLPath: String
    let fileType: String
    let cacheable: Bool
    let hash: String?
    let sourceMapPath: String?
    let sourceMapURLPath: String?
  }

  let version: String
  let cordovaCompatibilityVersion: String
  var entries: [Entry]

  init(fileURL: NSURL) throws {
    try self.init(data: try NSData(contentsOfURL: fileURL, options: []))
  }

  init(data: NSData) throws {
    let JSON: JSONObject
    do {
      JSON = try NSJSONSerialization.JSONObjectWithData(data, options: []) as! JSONObject
    } catch {
      throw WebAppError.InvalidAssetManifest(reason: "Error parsing asset manifest", underlyingError: error)
    }

    if let format = JSON["format"] as? String where format != "web-program-pre1" {
      throw WebAppError.InvalidAssetManifest(reason: "The asset manifest format is incompatible: \(format)", underlyingError: nil)
    }

    guard let version = JSON["version"] as? String else {
      throw WebAppError.InvalidAssetManifest(reason: "Asset manifest does not have a version", underlyingError: nil)
    }
    
    self.version = version
    
    guard let cordovaCompatibilityVersions = JSON["cordovaCompatibilityVersions"] as? JSONObject,
      let cordovaCompatibilityVersion = cordovaCompatibilityVersions["ios"] as? String else {
      throw WebAppError.InvalidAssetManifest(reason: "Asset manifest does not have a cordovaCompatibilityVersion", underlyingError: nil)
    }
    
    self.cordovaCompatibilityVersion = cordovaCompatibilityVersion
    
    let entriesJSON = JSON["manifest"] as? [JSONObject] ?? []
    entries = []
    for entryJSON in entriesJSON {
      if entryJSON["where"] as? String != "client" { continue }

      if let URLPath = entryJSON["url"] as? String,
        let filePath = entryJSON["path"] as? String,
        let fileType = entryJSON["type"] as? String,
        let hash = entryJSON["hash"] as? String,
        let cacheable = entryJSON["cacheable"] as? Bool {
          let sourceMapPath = entryJSON["sourceMap"] as? String
          let sourceMapURLPath = entryJSON["sourceMapUrl"] as? String

          let entry = Entry(filePath: filePath, URLPath: URLPath,
            fileType: fileType, cacheable: cacheable, hash: hash,
            sourceMapPath: sourceMapPath, sourceMapURLPath: sourceMapURLPath)
          entries.append(entry)
      }
    }
  }
}

File: src/android/AssetManifest.java
package com.meteor.webapp;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

final class AssetManifest {
    private static final String LOG_TAG = "MeteorWebApp";

    static final class Entry {
        final String filePath;
        final String urlPath;
        final String fileType;
        final boolean cacheable;
        final String hash;
        final String sourceMapFilePath;
        final String sourceMapUrlPath;

        Entry(String filePath, String urlPath, String fileType, boolean cacheable, String hash, String sourceMapFilePath, String sourceMapUrlPath) {
            this.filePath = filePath;
            this.urlPath = urlPath;
            this.fileType = fileType;
            this.cacheable = cacheable;
            this.hash = hash;
            this.sourceMapFilePath = sourceMapFilePath;
            this.sourceMapUrlPath = sourceMapUrlPath;
        }
    }

    final String version;
    final String cordovaCompatibilityVersion;
    final List<Entry> entries;

    public AssetManifest(String string) throws WebAppException {
        try {
            JSONObject json = new JSONObject(string);
            String format = json.optString("format", null);
            if (format != null && !format.equals("web-program-pre1")) {
                throw new WebAppException("The asset manifest format is incompatible: " + format);
            }

            try {
                version = json.getString("version");
            } catch (JSONException e) {
                throw new WebAppException("Asset manifest does not have a version", e);
            }

            try {
                JSONObject cordovaCompatibilityVersions = json.getJSONObject("cordovaCompatibilityVersions");
                cordovaCompatibilityVersion = cordovaCompatibilityVersions.getString("android");
            } catch (JSONException e) {
                throw new WebAppException("Asset manifest does not have a cordovaCompatibilityVersion", e);
            }

            JSONArray entriesJSON = json.getJSONArray("manifest");

            entries = new ArrayList<Entry>(entriesJSON.length());

            for (int i = 0; i < entriesJSON.length(); i++) {
                JSONObject entryJSON = entriesJSON.getJSONObject(i);

                if (!entryJSON.getString("where").equals("client")) continue;

                String filePath = entryJSON.getString("path");
                String urlPath = entryJSON.getString("url");

                String fileType = entryJSON.getString("type");
                boolean cacheable = entryJSON.getBoolean("cacheable");
                String hash = entryJSON.optString("hash", null);
                String sourceMapFilePath = entryJSON.optString("sourceMap", null);
                String sourceMapUrlPath = entryJSON.optString("sourceMapUrl", null);

                Entry entry = new Entry(filePath, urlPath, fileType, cacheable, hash, sourceMapFilePath, sourceMapUrlPath);
                entries.add(entry);
            }
        } catch (JSONException e) {
            throw new WebAppException("Error parsing asset manifest", e);
        }
    }
}
